mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp

```diff
@@ -558,22 +558,36 @@ class QuantizeLinearConverter final
     Type elementType = getShapedElementTy(input);
     Value inverseScale =
         createOpAndInfer<tosa::ReciprocalOp>(rewriter, loc, elementType, scale);
-    if (getShapedElementTy(inverseScale) != getShapedElementTy(input)) {
-      inverseScale = createCastOp(rewriter, loc, elementType, inverseScale);
-    }
     Value scaled = createOpAndInfer<tosa::MulOp>(
         rewriter, loc, elementType, input, inverseScale, /*shift=*/0);
 
-    Value shifted = scaled;
-    if (auto bias = op.getBias()) {
-      Value biasCast = createCastOp(rewriter, loc, elementType, bias);
-      shifted = createOpAndInfer<tosa::AddOp>(rewriter, loc, elementType,
-                                              scaled, biasCast);
-    }
-
     Type outputType = getShapedElementTy(output);
-    Value downCast = createCastOp(rewriter, loc, outputType, shifted);
-    rewriter.replaceOp(op, downCast);
+    // We cast the scaled value to the type of the bias, which might be larger
+    // than the final output type.
```

Comment:
What if this assumption doesn't satisfy and bias type is smaller than the final output type? If that's the case, we want to cast to scaled type and not clamp it.

BTW, is the motivation of this has to do with any existing migraphx IR/workload that has mixed type in bias?

Reply:
I don't know if there's an existing MIGraphX workload, but the MLIR quantization documentation specifically mentions int32 bias and int{8,16} outputs.

So we might as well support that.

But a bias smaller than the quantization type seems extremely unlikely

Reply:
In MIGraphX, the bias and quantized type must be the same.

Reply:
However, does this mean that adding the bias in is meant to wrap around?

Reply:
It should clamp.

Reply:
@pfultz2 But that means that I have to compute `scaled + bias` in a larger type, or else, with `scaled = 126` and `bias = 2`, we'd get `scaled + bias = -128`